tests/Authenticator.hs

```diff
@@ -0,0 +1,455 @@
+      --   error code equivalent to "NotAllowedError" and terminate the operation.
+      -- NOTE: We curently always succeed this step
+
+      -- 7. Once the authorization gesture has been completed and user consent
+      -- has been obtained, generate a new credential object:
+      -- 7.1. Let (publicKey, privateKey) be a new pair of cryptographic keys
+      -- using the combination of PublicKeyCredentialType and cryptographic
+      -- parameters represented by the first item in credTypesAndPubKeyAlgs that
+      -- is supported by this authenticator.
+      (publicKey, privateKey) <- newKeyPair $ M.pkcpAlg param
+
+      -- 7.2. Let userHandle be userEntity.id.
+      let userHandle = M.pkcueId userEntity
+
+      -- 7.4 If requireResidentKey is true or the authenticator chooses to
+      -- create a client-side discoverable public key credential source:
+
+      -- 7.4.1 Let credentialId be a new credential id.
+      -- NOTE: We need to have a CredentialId before we can construct the
+      -- credentialSource.
+      -- NOTE: We always choose to construct a clientside discoverable
+      -- credential, as this is allowed (See 7.4).
+      -- TODO: Use deterministic random number generator, and ensure that we
+      -- use a single random number generator across the entire library.
+      credentialId <- M.CredentialId <$> Random.getRandomBytes 16
+
+      -- 7.4.2. Set credentialSource.id to credentialId.
+      -- 7.3.  Let credentialSource be a new public key credential source with the fields:
+      let credentialSource =
+            PublicKeyCredentialSource
+              { pkcsId = credentialId,
+                pkcsPrivateKey = privateKey,
+                pkcsRpId = rpId,
+                pkcsUserHandle = Just userHandle
+              }
+
+      -- 7.4.3 Let credentials be this authenticator’s credentials map.
+      -- NOTE: we have aCredentials from the patternmatch
+      -- 7.4.4 Set credentials[(rpEntity.id, userHandle)] to credentialSource.
+
+      let credentials =
+            Map.insert (rpId, userHandle) credentialSource aCredentials
+
+      -- 8. If any error occurred while creating the new credential object,
+      -- return an error code equivalent to "UnknownError" and terminate the
+      -- operation.
+      -- NOTE: See above
+
+      -- 9. Let processedExtensions be the result of authenticator extension
+      -- processing for each supported extension identifier → authenticator
+      -- extension input in extensions.
+      -- NOTE: Extensions are unsupported
+
+      -- 10. If the authenticator supports a per credential signature counter,
+      -- allocate the counter, associate it with the new credential, and
+      -- initialize the counter value as zero.
+      -- NOTE: We return the updated signature counter and the supposed signatureCount.
+      -- The signatureCount will be 0 if Unsupported
+      let (signatureCount, signatureCounter) = initialiseCounter credentialId aSignatureCount
```

Comment:
This is rather confusing, I say let's rename this modules `SignatureCounter` to `AuthenticatorSignatureCounter` and then `aSignatureCount -> aSignatureCounter`. Also it makes more sense for `initialiseCounter` to be called `incrementCounter`

```suggestion
      let (signatureCounter, newASignatureCounter) = incrementCounter credentialId aSignatureCounter
```